File: leetcode/TwoSum1.js
var twoSum = function(nums, target) {
  var remainders = {};
  for(var i = 0; i < nums.length; i++) {
    var curr = nums[i];
    if(remainders[curr] !== undefined) {  // if the complement to the current number needed to equal the target exists
      return [remainders[curr], i];  // return the two indices whose respective values sum to the target
    } else {
      remainders[target - curr] = i;  // the remainder/difference you need to find in order to equal the target
    }
  }
  return [];
}

File: leetcode/TwoSum1.py
class Solution(object):
    def twoSum(self,nums,target):
        lookup = {}
        for i,num in enumerate(nums):
            if target-num in lookup: # difference we're looking for
                return [lookup[target-num],i] # indices of the two numbers adding up to target
            lookup[num] = i # otherwise, add key/value pair to lookup holding the current number and its index

if __name__ == '__main__':
    print Solution().twoSum((2,7,11,15),9)
    print Solution().twoSum((4,0,1),5)
